pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -242,6 +243,132 @@ type loadIdx struct {
 	chunk       int
 }
 
+type rangeLoadIdx struct {
+	offset uint32
+	estLen int64
+	// seriesEntry and rangeEntry are indices for where to put the result,
+	// they are not actual refs of series or chunks
+	seriesEntry int
+	rangeEntry  int
+}
+
+type bucketChunksRangesReader struct {
+	ctx    context.Context
+	block  *bucketBlock
+	toLoad [][]rangeLoadIdx
+}
+
+func newBucketChunksRangesReader(ctx context.Context, block *bucketBlock) *bucketChunksRangesReader {
+	return &bucketChunksRangesReader{
+		ctx:    ctx,
+		block:  block,
+		toLoad: make([][]rangeLoadIdx, len(block.chunkObjs)),
+	}
+}
+
+func (r *bucketChunksRangesReader) Close() error {
+	r.block.pendingReaders.Done()
+	return nil
+}
+
+// reset resets the ranges scheduled for loading. It does not release any loaded ranges.
+func (r *bucketChunksRangesReader) reset() {
+	for i := range r.toLoad {
+		r.toLoad[i] = r.toLoad[i][:0]
+	}
+}
+
+func (r *bucketChunksRangesReader) addLoadRange(g seriesChunkRefsRange, seriesEntry, rangeEntry int) error {
+	var (
+		seq = g.segmentFile
+		off = g.refs[0].segFileOffset
+	)
+	if seq >= uint32(len(r.toLoad)) {
+		return errors.Errorf("reference sequence %d out of range", seq)
+	}
+
+	var totalLen int64
+	for _, c := range g.refs {
+		totalLen += int64(c.length)
```

Comment:
I'm not convinced about this calculation. The range length should be "(offset of the last chunk - offset of the first chunk) + length of the last chunk"?

Reply:
If we do that, then we're bringing some length estimation logic into `bucketChunksRangesReader`. 

There is some logic in `loadingSeiresChunkRefsSetIterator` about clamping the length of the chunk if it's too big (in case chunks are scattered around the segment file). If we do the ref math here as well we will override that logic for example and the size of the range could become very large.

> I'm not convinced about this calculation

any particular donwsides you see about this?

Reply:
I think I'm getting tricked by the meaning of `c.length`. When I did this review I thought it was the chunk length (in the sense of the actual chunk data length) but reading the rest of the code I think it actually is the length of the following struct, right?

```
		// ┌───────────────┬───────────────────┬──────────────┬────────────────┐
		// │ len <uvarint> │ encoding <1 byte> │ data <bytes> │ CRC32 <4 byte> │
		// └───────────────┴───────────────────┴──────────────┴────────────────┘
```

If I'm correct, then the current logic LGTM, but I would:
1. Add a comment to explain why we're doing this math here
2. Improve the field comment of `c.length` to clarify what that length is

Reply:
> I think I'm getting tricked by the meaning of c.length. When I did this review I thought it was the chunk length (in the sense of the actual chunk data length) but reading the rest of the code I think it actually is the length of the following struct, right?

that's right. The length is the total length occupied in the segment file, not only the data field of a chunk.